Terror Spider Fixes & Tweaks

Kyep opened the pull request:
Fixes #11782 
Fixes #11506 

:cl: Kyep
fix: Terror spider away mission is no longer unplayable due to lack of spiders spawning
fix: Terror spiders no longer smash random things for no reason
fix: Gray terror spiders now use the correct web type
tweak: Tweaked gateway terror spider behavior a bit to account for the fact their AI is now only active when living players are in the gateway
refactor: Cleaned up some terror spider code
fix: White terror spiders (and their webs) are no longer able to infect slimes.
fix: Terror spiders in general no longer allow you to control them if you've rendered yourself unrespawnable
tweak: Reworked gray terrors. Previously they (according to class role, and AI coding) hid in vents and ambushed people who got close or walked into vents. Players did not play this way, though, so now they've been changed to simply be stealthed (like xeno hunters) whenever they're on a web. Their AI has also been dramatically simplified. To balance this ability, they've lost the fast health regen they used to have.
fix: Fixed case of terror spider attack animations not working (AI spider attacking light, prince using special ability)
fix: Fixed white terror infections not properly causing confusion
/:cl:

Emanthealmighty commented:
Could this be applied to giant spiders too?

Kyep commented:
Yes, but if we're wanting to apply it to other mobs, I might change the approach. Instead of overriding the proc on /terrorspider, instead I'd just have terrorspider + nurse regular spider + whatever else have a var like 'aiZLevelCheck' or something which is set to TRUE for those mobs that are always active whenever a player is on their zlevel.

Emanthealmighty commented:
I'd say that having a general var like the one you described would be ideal.

Kyep commented:
@Emanthealmighty changed, it now affects nurse spiders of the giant spiders (but not the hunter/guard giant spider)

Emanthealmighty commented:
> @Emanthealmighty it now affects giant nurse spiders (the green-eyed giant spiders that lay eggs) too. It still doesn't affect the hunter/guard giant spider, as I don't think the effects on them would be too useful. There may be other bugs with giant spiders that stop them wrapping things, but at least with this fix, it won't be the zlevel/distance stuff.
> 
> I'm still not happy with how sluggish gateway terror spiders are to respond when a player gets in range of them... but I am not sure how to change that (or even if it is changable without having them scan much more often for hostiles).
> 
> Still, they're once again populating the mission, and they do respond (eventually) when you get in range of them. So, the mission is playable again.

I think making them scan more often would be a solution. Before the gateway mission broke, I noticed they don't even wrap up bodies that often, while surprisingly, giant spiders did.

Kyep commented:
@Emanthealmighty Making them scan more often would be a solution, one that I might implement in future. Wrapping bodies is meant to be a low-frequency behavior as they get little benefit from it and there are unlikely to be bodies in the gateway for them to wrap.

Kyep commented:
By 'in future' I mean 'in another PR'. I think this PR is ready as-is.

Kyep commented:
Test

Kyep commented:
This PR has been rewritten. Major changes:
- Much more focus on code quality (it now includes some related refactors and cleanup, it no longer overrides consider_wakeup, etc)
- It now works better for giant spiders as well
Description updated.

Kyep commented:
More fixes added:
fix: White terror spiders (and their webs) are no longer able to infect slimes.
fix: Terror spiders in general no longer allow you to control them if you've rendered yourself unrespawnable

Kyep commented:
Added:
tweak: Reworked gray terrors. Previously they (according to class role, and AI coding) hid in vents and ambushed people who got close or walked into vents. Players did not play this way, though, so now they've been changed to simply be stealthed (like xeno hunters) whenever they're on a web. Their AI has also been dramatically simplified.

Fox-McCloud commented on code/modules/mob/living/simple_animal/hostile/terror_spiders/terror_ai.dm:
```diff
@@ -92,10 +90,18 @@
+/mob/living/simple_animal/hostile/poison/terror_spider/handle_automated_movement()
+	// Putting the main terror spider AI code in handle_automated_movement() rather than handle_automated_action() ensures it will still run when the spider AIStatus == AI_IDLE
+	// This is necessary for the terror spiders in the away mission to work properly.
+	if(stat || ckey || AIStatus != AI_IDLE)
```
there's no way that `handle_automated_movement` gets called on something with that's dead or has a ckey.

Kyep replied:
Updated to remove those two checks.

Kyep commented:
Updated to fix merge conflict and remove stuff related to giant spiders that was changed in Fox's PR.

Kyep commented:
pdated to fix merge conflict and remove stuff related to giant spiders that was changed in Fox's PR.
